test/micro/org/openjdk/bench/java/security/HSSLMS.java

```diff
@@ -0,0 +1,1753 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package org.openjdk.bench.java.security;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.io.IOException;
+import java.util.HexFormat;
+import java.security.ProviderException;
+import java.security.PublicKey;
+import java.security.Security;
+import java.security.Provider;
+import java.security.KeyFactory;
+import java.security.Signature;
+import java.util.concurrent.TimeUnit;
+
+import sun.security.util.RawKeySpec;
+
+/**
+ * Benchmark measuring HSS/LMS
+ */
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@State(Scope.Thread)
+@Warmup(iterations = 5, time = 1)
+@Measurement(iterations = 5, time = 1)
+@Fork(value = 3, jvmArgsAppend = {"--add-exports", "java.base/sun.security.util=ALL-UNNAMED"})
+
+// Tests 1-2 are from RFC 8554, Appendix F.
+
+// Tests 3-10 were generated with parameter sets mentioned in
+// RFC 8554 section 6.4; two with W=8 and six with W=4.
+
+public class HSSLMS {
+    Provider p;
+
+    @Benchmark
+    public void test01_RFC_8554() throws Exception {
+        // RFC 8554 Test Case 1
+        var pk = decode("""
```

Comment:
If we are primarily interested in measuring the performance of the HSS/LMS verification, the decode parts should really be left out of that calculation and moved to `setup()` methods. Consider having separate inner/static subclasses for each test, and moving the decoding code to the `setup()` methods in each of those subclasses. See `test/micro/org/openjdk/bench/java/security/Signatures.java` for an example.

Reply:
Fixed. It made a measurable difference.